## Analysis of Used Car listings
This project aims to clean and analyze a  used car dataset obtained from eBay Kleinanzeigen, a classifieds section of the German eBay website.

In [1]:
import pandas as pd

autos = pd.read_csv('autos.csv', encoding='Latin-1')
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [2]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [3]:
# autos_columns = autos.columns.copy()
# autos_columns

autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [4]:
autos.rename(columns = {'dateCrawled': 'date_crawled',
                     'offerType': 'offer_type',
                     'vehicleType': 'vehicle_type',
                     'yearOfRegistration': 'registration_year',
                     'monthOfRegistration': 'registration_month',
                     'fuelType': 'fuel_type',
                     'notRepairedDamage': 'unrepaired_damage',
                     'dateCreated': 'ad_created',
                     'nrOfPictures': 'nr_of_pictures',
                     'postalCode': 'postal_code',
                     'lastSeen': 'last_seen'}, inplace=True)
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [7]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-27 22:55:05  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Looking at the data, we have some cleaning tasks to do. Columns that dont have much uniqueness about them have to be dropped, columns that are in text format instead of numeric need to be converted.

In [9]:
autos.drop(['seller', 'offer_type', 'nr_of_pictures'], axis=1)

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

         price   abtest vehicle_type  registration_year    gearbox  powerPS  \
0       $5,000  control          bus               2004    manuell      158   
1       $8,500  control    limousine               1997  automatik      286   
2       $8,990     test    limousine               2009    manuell      102   
3       $4,350  control   kleinwagen               2007  automatik       71   
4       $1,350     test        kombi               2003    manuell        0   
...        ...      ...          ...                ...        ...      ...   
49995  $24,900  control    limousine               2011  automatik      239   
49996   $1,980  control       cabrio               1996    manuell       75   
49997  $13,200     test       cabrio               2014  automatik       69   
49998  $22,900  control        kombi               2013    manuell      150   
49999   $1,250  control    limousine               1996    manuell      101   

        model   odometer  registration_month fuel_type       brand  \
0      andere  150,000km                   3       lpg     peugeot   
1         7er  150,000km                   6    benzin         bmw   
2        golf   70,000km                   7    benzin  volkswagen   
3      fortwo   70,000km                   6    benzin       smart   
4       focus  150,000km                   7    benzin        ford   
...       ...        ...                 ...       ...         ...   
49995      q5  100,000km                   1    diesel        audi   
49996   astra  150,000km                   5    benzin        opel   
49997     500    5,000km                  11    benzin        fiat   
49998      a3   40,000km                  11    diesel        audi   
49999  vectra  150,000km                   1    benzin        opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
0                  nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                  nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                  nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3                  nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4                  nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50  
...                 ...                  ...          ...                  ...  
49995              nein  2016-03-27 00:00:00        82131  2016-04-01 13:47:40  
49996              nein  2016-03-28 00:00:00        44807  2016-04-02 14:18:02  
49997              nein  2016-04-02 00:00:00        73430  2016-04-04 11:47:27  
49998              nein  2016-03-08 00:00:00        35683  2016-04-05 16:45:07  
49999              nein  2016-03-13 00:00:00        45897  2016-04-06 21:18:48  

[50000 rows x 17 columns]

In [10]:
autos['price']  = autos['price'].str.replace('$', '').str.replace(',', '').astype(int)

In [12]:
autos['odometer'] = autos['odometer'].str.replace(',', '').str.replace('km', '').astype(int)

In [35]:
autos.rename(columns = {'odometer':'odometer_km'}, inplace=True)

## Exploring the Price and Odometer columns

In [36]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [37]:
autos['price'].value_counts().sort_values(ascending=False).head(20)

0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price, dtype: int64

In [38]:
autos['price'].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [39]:
autos[autos['price'].between(1, 250000)]

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type  price   abtest vehicle_type  registration_year  \
0      privat    Angebot   5000  control          bus               2004   
1      privat    Angebot   8500  control    limousine               1997   
2      privat    Angebot   8990     test    limousine               2009   
3      privat    Angebot   4350  control   kleinwagen               2007   
4      privat    Angebot   1350     test        kombi               2003   
...       ...        ...    ...      ...          ...                ...   
49995  privat    Angebot  24900  control    limousine               2011   
49996  privat    Angebot   1980  control       cabrio               1996   
49997  privat    Angebot  13200     test       cabrio               2014   
49998  privat    Angebot  22900  control        kombi               2013   
49999  privat    Angebot   1250  control    limousine               1996   

         gearbox  powerPS   model  odometer_km  registration_month fuel_type  \
0        manuell      158  andere       150000                   3       lpg   
1      automatik      286     7er       150000                   6    benzin   
2        manuell      102    golf        70000                   7    benzin   
3      automatik       71  fortwo        70000                   6    benzin   
4        manuell        0   focus       150000                   7    benzin   
...          ...      ...     ...          ...                 ...       ...   
49995  automatik      239      q5       100000                   1    diesel   
49996    manuell       75   astra       150000                   5    benzin   
49997  automatik       69     500         5000                  11    benzin   
49998    manuell      150      a3        40000                  11    diesel   
49999    manuell      101  vectra       150000                   1    benzin   

            brand unrepaired_damage           ad_created  nr_of_pictures  \
0         peugeot              nein  2016-03-26 00:00:00               0   
1             bmw              nein  2016-04-04 00:00:00               0   
2      volkswagen              nein  2016-03-26 00:00:00               0   
3           smart              nein  2016-03-12 00:00:00               0   
4            ford              nein  2016-04-01 00:00:00               0   
...           ...               ...                  ...             ...   
49995        audi              nein  2016-03-27 00:00:00               0   
49996        opel              nein  2016-03-28 00:00:00               0   
49997        fiat              nein  2016-04-02 00:00:00               0   
49998        audi              nein  2016-03-08 00:00:00               0   
49999        opel              nein  2016-03-13 00:00:00               0   

       postal_code            last_seen  
0            79588  2016-04-06 06:45:54  
1            71034  2016-04-06 14:45:08  
2            35394  2016-04-06 20:15:37  
3            33729  2016-03-15 03:16:28  
4            39218  2016-04-01 14:38:50  
..

In [40]:
autos['odometer_km'].unique().shape

(13,)

In [41]:
autos['odometer_km'].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [42]:
autos[['date_crawled', 'ad_created', 'last_seen']].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [43]:
autos[['date_crawled', 'ad_created', 'last_seen']].describe()

date_crawled           ad_created            last_seen
count                 50000                50000                50000
unique                48213                   76                39481
top     2016-03-27 22:55:05  2016-04-03 00:00:00  2016-04-07 06:17:27
freq                      3                 1946                    8

In [53]:
autos['date_crawled'].value_counts(normalize=True, dropna=False)[:15].sort_index()

2016-03-05 16:57:05    0.00006
2016-03-08 10:40:35    0.00006
2016-03-09 11:54:38    0.00006
2016-03-12 16:06:22    0.00006
2016-03-14 20:50:02    0.00006
2016-03-16 21:50:53    0.00006
2016-03-19 17:36:18    0.00006
2016-03-22 09:51:06    0.00006
2016-03-23 18:39:34    0.00006
2016-03-27 22:55:05    0.00006
2016-03-29 23:42:13    0.00006
2016-03-30 17:37:35    0.00006
2016-03-30 19:48:02    0.00006
2016-04-02 11:37:04    0.00006
2016-04-02 15:49:30    0.00006
Name: date_crawled, dtype: float64

In [54]:
autos['ad_created'].value_counts(normalize=True, dropna=False)[:15].sort_index()

2016-03-07 00:00:00    0.03474
2016-03-08 00:00:00    0.03334
2016-03-12 00:00:00    0.03662
2016-03-14 00:00:00    0.03522
2016-03-15 00:00:00    0.03374
2016-03-19 00:00:00    0.03384
2016-03-20 00:00:00    0.03786
2016-03-21 00:00:00    0.03772
2016-03-28 00:00:00    0.03496
2016-03-29 00:00:00    0.03414
2016-03-30 00:00:00    0.03344
2016-04-01 00:00:00    0.03380
2016-04-02 00:00:00    0.03508
2016-04-03 00:00:00    0.03892
2016-04-04 00:00:00    0.03688
Name: ad_created, dtype: float64

In [55]:
autos['last_seen'].value_counts(normalize=True, dropna=False)[:15].sort_index()

2016-04-05 16:44:47    0.00012
2016-04-06 01:16:01    0.00012
2016-04-06 06:17:24    0.00014
2016-04-06 07:46:11    0.00012
2016-04-06 08:44:19    0.00012
2016-04-06 10:17:12    0.00012
2016-04-06 13:17:03    0.00012
2016-04-06 14:44:55    0.00012
2016-04-06 15:45:50    0.00012
2016-04-06 20:48:27    0.00012
2016-04-06 21:17:51    0.00014
2016-04-07 03:16:17    0.00014
2016-04-07 04:46:51    0.00012
2016-04-07 05:16:17    0.00012
2016-04-07 06:17:27    0.00016
Name: last_seen, dtype: float64

In [56]:
autos['registration_year'].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

### The maximum year is 9999 and the minimum year is 1000, this isnt acceptable so we have to further clean the year column to a  more realistic timeline it 1900 - 2016, 2016 being the most current year in the dataset

In [63]:
autos = autos[autos['registration_year'].between(1900, 2016)]

In [64]:
autos['registration_year'].describe()

count    48028.00000
mean      2002.80351
std          7.31085
min       1910.00000
25%       1999.00000
50%       2003.00000
75%       2008.00000
max       2016.00000
Name: registration_year, dtype: float64

In [74]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [75]:
#autos['brand'].value_counts(normalize=True)
top_brand = autos['brand'].value_counts(normalize=True)
top_brands = top_brand[top_brand > 0.05].index
top_brands


Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [88]:
top_brands_mean_price = {}

for brand in top_brands:
    brand_only = autos[autos['brand'] == brand]
    mean_price = brand_only['price'].mean()
    top_brands_mean_price[brand] = int(mean_price)
    
top_brands_mean_price

{'volkswagen': 6516,
 'bmw': 8334,
 'opel': 5252,
 'mercedes_benz': 30317,
 'audi': 9093,
 'ford': 7263}

In [90]:
mileage = {}

for brand in top_brands:
    brand_only = autos[autos['brand'] == brand]
    miles_mean = brand_only['odometer_km'].mean()
    mileage[brand] = int(miles_mean)
    
mileage

{'volkswagen': 128730,
 'bmw': 132434,
 'opel': 129227,
 'mercedes_benz': 130860,
 'audi': 129287,
 'ford': 124046}

In [91]:
m_prices = pd.Series(top_brands_mean_price).sort_values(ascending=False)
m_prices

mercedes_benz    30317
audi              9093
bmw               8334
ford              7263
volkswagen        6516
opel              5252
dtype: int64

In [92]:
m_mileage = pd.Series(mileage).sort_values(ascending=False)
m_mileage

bmw              132434
mercedes_benz    130860
audi             129287
opel             129227
volkswagen       128730
ford             124046
dtype: int64

In [98]:
d_price = pd.DataFrame(m_prices, columns=['Average Price'])
d_price

Average Price
mercedes_benz          30317
audi                    9093
bmw                     8334
ford                    7263
volkswagen              6516
opel                    5252

In [99]:
d_mileage = pd.DataFrame(m_mileage, columns=['Average Mileage'])
d_mileage

Average Mileage
bmw                     132434
mercedes_benz           130860
audi                    129287
opel                    129227
volkswagen              128730
ford                    124046